qiskit_optimization/algorithms/qrao/quantum_random_access_optimizer.py

```diff
@@ -0,0 +1,304 @@
+from qiskit_optimization.problems import QuadraticProgram, Variable
+
+from .quantum_random_access_encoding import QuantumRandomAccessEncoding
+from .rounding_common import RoundingContext, RoundingResult, RoundingScheme
+from .semideterministic_rounding import SemideterministicRounding
+
+
+class QuantumRandomAccessOptimizationResult(OptimizationResult):
+    """Result of Quantum Random Access Optimization procedure."""
+
+    def __init__(
+        self,
+        *,
+        x: Optional[Union[List[float], np.ndarray]],
+        fval: Optional[float],
+        variables: List[Variable],
+        status: OptimizationResultStatus,
+        samples: Optional[List[SolutionSample]],
+        encoding: QuantumRandomAccessEncoding,
+        relaxed_fval: float,
+        relaxed_result: MinimumEigensolverResult,
+        rounding_result: RoundingResult,
+    ) -> None:
+        """
+        Args:
+            x: The optimal value found by ``MinimumEigensolver``.
+            fval: The optimal function value.
+            variables: The list of variables of the optimization problem.
+            status: The termination status of the optimization algorithm.
+            samples: The list of ``SolutionSample`` obtained from the optimization algorithm.
+            encoding: The encoding used for the optimization.
+            relaxed_fval: The optimal function value of the relaxed problem.
+            relaxed_result: The result obtained from the underlying minimum eigensolver.
+            rounding_result: The rounding result.
+        """
+        super().__init__(
+            x=x,
+            fval=fval,
+            variables=variables,
+            status=status,
+            raw_results=None,
+            samples=samples,
+        )
+        self._encoding = encoding
+        self._relaxed_fval = relaxed_fval
+        self._relaxed_result = relaxed_result
+        self._rounding_result = rounding_result
+
+    @property
+    def encoding(self) -> QuantumRandomAccessEncoding:
+        """The encoding used for the optimization."""
+        return self._encoding
+
+    @property
+    def relaxed_fval(self) -> float:
+        """The optimal function value of the relaxed problem."""
+        return self._relaxed_fval
+
+    @property
+    def relaxed_result(
+        self,
+    ) -> MinimumEigensolverResult:
+        """The result obtained from the underlying minimum eigensolver."""
+        return self._relaxed_result
+
+    @property
+    def rounding_result(self) -> RoundingResult:
+        """The rounding result."""
+        return self._rounding_result
+
+
+class QuantumRandomAccessOptimizer(OptimizationAlgorithm):
+    """Quantum Random Access Optimizer class."""
+
+    def __init__(
+        self,
+        min_eigen_solver: MinimumEigensolver,
+        penalty: Optional[float] = None,
+        max_vars_per_qubit: int = 3,
+        rounding_scheme: Optional[RoundingScheme] = None,
```

Comment:
`float | None` and `Rounding | None` rather than Optional. I will not comment any further - its probably simpler to search the code for Optional and change them....